google-cloud-pubsub/lib/google/cloud/pubsub/subscriber.rb

```diff
@@ -13,417 +13,565 @@
       #
       # @example
       #   require "google/cloud/pubsub"
       #
       #   pubsub = Google::Cloud::PubSub.new
       #
-      #   sub = pubsub.subscription "my-topic-sub"
-      #
-      #   subscriber = sub.listen do |received_message|
+      #   subscriber = pubsub.subscriber "my-topic-sub"
+      #   listener = subscriber.listen do |received_message|
       #     # process message
       #     received_message.acknowledge!
       #   end
       #
-      #   # Start background threads that will call the block passed to listen.
-      #   subscriber.start
-      #
-      #   # Shut down the subscriber when ready to stop receiving messages.
-      #   subscriber.stop!
+      #   # Handle exceptions from listener
+      #   listener.on_error do |exception|
+      #      puts "Exception: #{exception.class} #{exception.message}"
+      #   end
       #
-      # @attr_reader [String] subscription_name The name of the subscription the
-      #   messages are pulled from.
-      # @attr_reader [Proc] callback The procedure that will handle the messages
-      #   received from the subscription.
-      # @attr_reader [Numeric] deadline The default number of seconds the stream
-      #   will hold received messages before modifying the message's ack
-      #   deadline. The minimum is 10, the maximum is 600. Default is 60.
-      # @attr_reader [Boolean] message_ordering Whether message ordering has
-      #   been enabled.
-      # @attr_reader [Integer] streams The number of concurrent streams to open
-      #   to pull messages from the subscription. Default is 2.
-      # @attr_reader [Integer] callback_threads The number of threads used to
-      #   handle the received messages. Default is 8.
-      # @attr_reader [Integer] push_threads The number of threads to handle
-      #   acknowledgement ({ReceivedMessage#ack!}) and delay messages
-      #   ({ReceivedMessage#nack!}, {ReceivedMessage#modify_ack_deadline!}).
-      #   Default is 4.
+      #   # Gracefully shut down the subscriber
+      #   at_exit do
+      #     listener.stop!
+      #   end
       #
+      #   # Start background threads that will call the block passed to listen.
+      #   listener.start
+      #   sleep
       class Subscriber
-        include MonitorMixin
-
-        attr_reader :subscription_name
-        attr_reader :callback
-        attr_reader :deadline
-        attr_reader :streams
-        attr_reader :message_ordering
-        attr_reader :callback_threads
-        attr_reader :push_threads
-
         ##
-        # @private Implementation attributes.
-        attr_reader :stream_pool, :thread_pool, :buffer, :service
+        # @private The Service object.
+        attr_accessor :service
 
         ##
-        # @private Implementation attributes.
-        attr_accessor :exactly_once_delivery_enabled
+        # @private The gRPC Google::Cloud::PubSub::V1::Subscription object.
+        attr_accessor :grpc
```

Comment:
Since we're now telling users to use the GAPIC for admin operations, it would probably be a good idea to make the Subscription resource available and public so they can use it in admin calls.

NB:

* We shouldn't just make this attr public since `grpc` is a terrible name for it. (I like something like `subscription_resource`.)
* Remember that the `@grpc` field could be nil if this object is a reference. May need to call `ensure_grpc!` when implementing a public accessor.